mirascope/mcp/client.py

```diff
@@ -0,0 +1,280 @@
+
+        if field_name in required_fields:
+            default = Field(..., description=field_schema.get("description"))
+            annotation = field_type
+        else:
+            default = Field(None, description=field_schema.get("description"))
+            annotation = field_type | None
+
+        fields[field_name] = (annotation, default)
+
+    return create_model(
+        snake_to_pascal(tool.name), __config__=ConfigDict(extra="allow"), **fields
+    )
+
+
+def _create_model_from_prompt(prompt: Prompt) -> type[BaseModel]:
+    if not prompt.arguments:
+        return create_model(prompt.name, __base__=MCPClientTool)
+
+    fields = {}
+    for arg in prompt.arguments:
+        if arg.required:
+            default = Field(..., description=arg.description)
+            annotation = str
+        else:
+            default = Field(None, description=arg.description)
+            annotation = str | None
+
+        fields[arg.name] = annotation, default
+
+    return create_model(prompt.name, __base__=MCPClientTool, **fields)
+
+
+class MCPClient:
+    def __init__(self, session: ClientSession) -> None:
+        self.session: ClientSession = session
+
+    async def list_resources(self) -> list[Resource]:
+        result = await self.session.list_resources()
+        return result.resources
+
+    async def read_resource(
+        self, uri: str | AnyUrl
+    ) -> list[str | BlobResourceContents]:
+        result = await self.session.read_resource(
+            uri if isinstance(uri, AnyUrl) else uri
+        )
+        parsed_results = []
+        for content in result.contents:
+            if isinstance(result, BlobResourceContents):
+                parsed_results.append(result)
+            else:
+                parsed_results.append(content.text)
+        return parsed_results
+
+    async def list_prompts(self) -> list[Prompt]:
+        result = await self.session.list_prompts()
+        return result.prompts
+
+    async def get_prompt_template(
+        self, name: str
+    ) -> Callable[..., Awaitable[list[BaseMessageParam]]]:
+        async def async_prompt(**kwargs: str) -> list[BaseMessageParam]:
+            result = await self.session.get_prompt(name, kwargs)
+
+            return [
+                _convert_prompt_message_to_base_message_params(prompt_message)
+                for prompt_message in result.messages
+            ]
+
+        return async_prompt
+
+    async def list_tools(self) -> list[type[MCPClientTool]]:
+        list_tool_result = await self.session.list_tools()
+
+        converted_tools = []
+        for tool in list_tool_result.tools:
+            model = create_model_from_tool(tool)
+
+            class CustomMCPTool(MCPClientTool):
+                __call_tool__ = self.session.call_tool
+                __custom_name__ = tool.name
+
+            converted_tools.append(
+                convert_base_model_to_base_tool(model, CustomMCPTool)
+            )
```

Comment:
This is the best I can come up with, can you think of any other good approaches?

Reply:
why not just push the client tool functionality into the server tool so that the tool returned by the server is just ready to go?

Reply:
@willbakst 
Sorry, I’m a bit confused. Since the data returned by the server is already predetermined, I believe we need to construct the tool on the client side based on the interface definition provided by the server’s tool.